src/main/java/jewellery/inventory/service/ProductService.java

```diff
@@ -0,0 +1,230 @@
+
+        Product product = setFieldsToNewProduct(productRequestDto, user, resourcesInProduct);
+        productRepository.save(product);
+
+        setContentProduct(productRequestDto, product);
+        setProductToResourcesInProduct(resourcesInProduct, product);
+
+        return productMapper.mapToProductResponseDto(product);
+    }
+
+    public List<ProductResponseDto> getAllProducts() {
+        List<Product> products = productRepository.findAll();
+        return products.stream().map(productMapper::mapToProductResponseDto).toList();
+    }
+
+    public ProductResponseDto getProduct(UUID id) {
+        Product product = productRepository.findById(id)
+                .orElseThrow(() -> new ProductNotFoundException(id));
+
+        return productMapper.mapToProductResponseDto(product);
+    }
+
+    public void deleteProduct(UUID id) {
+
+        Product product = productRepository.findById(id).orElseThrow(
+                () -> new ProductNotFoundException(id));
+
+        throwExceptionIfProductIsSold(id, product);
+        throwExceptionIfProductIsPartOfAnotherProduct(id, product);
+
+        moveResourceInProductToResourceInUser(product);
+        disassembleProductContent(product);
+
+        productRepository.deleteById(id);
+    }
+
+    private void throwExceptionIfProductIsPartOfAnotherProduct(UUID id, Product product) {
+        if (product.getContentOf() != null) {
+            throw new ProductIsContentException(id);
+        }
+    }
+
+    private void throwExceptionIfProductIsSold(UUID id, Product product) {
+        if (product.isSold()) {
+            throw new ProductIsSoldException(id);
+        }
+    }
+
+    private void disassembleProductContent(Product product) {
+        if (product.getProductsContent() != null) {
+            product.getProductsContent().forEach(content -> {
+                content.setContentOf(null);
+                productRepository.save(content);
+            });
+
+            product.setProductsContent(new ArrayList<>());
+            productRepository.save(product);
+        }
+    }
+
+    private void moveResourceInProductToResourceInUser(Product product) {
+        List<ResourceInProduct> resourcesInProduct = product.getResourcesContent();
+        User owner = product.getOwner();
+
+        List<ResourceInUser> resultResourcesInUser = new ArrayList<>();
+
+        resourcesInProduct.forEach(resourceInProduct -> {
+            Resource resource = resourceInProduct.getResource();
+            ResourceInUser resourceInUser = moveResourcesToUserWhenDisassembleProduct(owner, resourceInProduct, resource);
+
+            resultResourcesInUser.add(resourceInUser);
+
+            if (resourceInProduct.getProduct() == null) {
+                resourceInProductRepository.deleteById(resourceInProduct.getId());
+            }
+        });
+
+        resourceInUserRepository.saveAll(resultResourcesInUser);
```

Comment:
Instead of this, can we try to utilise ResourceInUserService::addResourceToUser? It should be nice to reuse functionality instead of writing it anew. We'd need to map the ResourceInProduct to ResourceInUserRequest, but that should be easy enough.